### PR TITLE
FIX Validates target in cosine_embedding

### LINKED ISSUE #53030: Computation for CosineEmbeddingLoss is incorrect when target has a batch dim
## 🐛 Bug

Identified in #52732

## To Reproduce

As demonstrated in the linked issue, passing a target of shape `(N, 1)` produces a different loss value than passing a target of shape `N`:

```python
import torch
import torch.nn as nn

torch.manual_seed(0)

a = torch.randn(16, 5)
b = torch.randn(16, 5)
labels = torch.empty(16).bernoulli_().mul_(2).sub_(1)

loss_fn = nn.CosineEmbeddingLoss()

print(loss_fn(a, b, labels))  # 0.4538
print(loss_fn(a, b, labels.unsqueeze(1)))  # 0.4810
```

## Expected behavior

Resultant loss should not be different for target of shape `N` vs. target of shape `(N, 1)`. Either the computation should be corrected or an error should be thrown for the latter case.

## Additional Context
The problem is due to the usage of `torch.where` within the loss computation. Its broadcasting behavior results in a different value when the target has the extra dim.

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @albanD @mruberry

### PR DESCRIPTION
**Pull request by thomasjpfan**
Fixes #53030

This PR validates the target for `cosine_embedding_loss`. This is consistent with how `cross_entropy` handles non 1d targets:

```py
import torch
import torch.nn.functional as F

input = torch.randn(3, 5, requires_grad=True)
target = torch.randint(5, (3, 1))

# Raises RuntimeError
loss = F.cross_entropy(input, target)
```

### CODE REVIEW

**Review by jbschlosser — APPROVED** [review]
LGTM! Thanks for the thorough fix & accompanying test :)
